jxl/src/var_dct/dct.rs

```diff
@@ -0,0 +1,255 @@
+            }
+            fn multiply_and_add(coeff: &[[f32; SZ]], out: &mut [[f32; SZ]]) {
+                for i in 0..($nhalf) {
+                    for j in 0..SZ {
+                        let mul = WcMultipliers::<$n>::K_MULTIPLIERS[i];
+                        let in1 = coeff[i][j];
+                        let in2 = coeff[$nhalf + i][j];
+                        out[i][j] = mul * in2 + in1;
+                        out[($n - i - 1)][j] = -mul * in2 + in1;
+                    }
+                }
+            }
+        }
+
+        impl IDCT1D for IDCT1DImpl<$n> {
+            fn do_idct<const COLUMNS: usize>(data: &mut [[f32; COLUMNS]]) {
+                const { assert!($nhalf * 2 == $n) }
+
+                // We assume `data` is arranged as a 4xCOLUMNS matrix in a flat array.
+
+                let mut tmp = [[0.0f32; COLUMNS]; $n]; // Temporary buffer
+
+                // 1. ForwardEvenOdd
+                CoeffBundle::<$n, COLUMNS>::forward_even_odd(data, &mut tmp);
+                // 2. First Recursive Call (IDCT1DImpl::do_idct)
+                // first half
+                IDCT1DImpl::<$nhalf>::do_idct::<COLUMNS>(&mut tmp[0..$nhalf]);
+                // 3. BTranspose.
+                // only the second half
+                CoeffBundle::<$nhalf, COLUMNS>::b_transpose(&mut tmp[$nhalf..$n]);
+                // 4. Second Recursive Call (IDCT1DImpl::do_idct)
+                // second half
+                IDCT1DImpl::<$nhalf>::do_idct::<COLUMNS>(&mut tmp[$nhalf..$n]);
+                // 5. MultiplyAndAdd.
+                CoeffBundle::<$n, COLUMNS>::multiply_and_add(&tmp, data);
+            }
+        }
+    };
+}
+define_idct_1d!(4, 2);
+define_idct_1d!(8, 4);
+define_idct_1d!(16, 8);
+define_idct_1d!(32, 16);
+define_idct_1d!(64, 32);
+define_idct_1d!(128, 64);
+define_idct_1d!(256, 128);
+
+fn transpose<const N: usize, const M: usize>(_data: &mut [f32]) {
+    todo!();
+}
+
+fn idct2d<const ROWS: usize, const COLS: usize>(_data: &mut [f32])
+where
+    IDCT1DImpl<ROWS>: IDCT1D,
+    IDCT1DImpl<COLS>: IDCT1D,
+{
+    todo!();
+    // TODO(firsching): Add something like
+    // IDCT1DImpl::<ROWS>::do_idct::<COLS>(data);
+    // transpose::<ROWS, COLS>(data);
+    // IDCT1DImpl::<COLS>::do_idct::<ROWS>(data);
+}
+
+#[cfg(test)]
+mod tests {
+    use std::array;
+
+    use crate::{
+        util::test::assert_almost_eq,
+        var_dct::{
+            dct::{IDCT1DImpl, IDCT1D},
+            dct_slow::idct1d,
+        },
+    };
+    macro_rules! test_idct1d_eq_slow_n {
+        ($test_name:ident, $n_val:expr, $tolerance:expr) => {
+            #[test]
+            fn $test_name() {
+                const N: usize = $n_val;
+                const M: usize = 1;
+                const NM: usize = N * M;
+
+                // Generate input data
+                let input_f64_vec: Vec<f64> = (1..=N).map(|i| i as f64).collect();
+                let input_f64: [f64; NM] = input_f64_vec
+                    .try_into()
```

Comment:
it's fine to do `to_vec()` in tests and not handle allocation errors

Reply:
huh, I don't quite follow, what do you suggest here exactly?

Reply:
ignore me, I misread :-)